### DISCUSSION TITLE
Memory leak while copying VImage before drawing

### DISCUSSION THREAD

**Discussion by MathemanFlo** [libvips/libvips#2733] [Q&A] [answered]
**Describe the bug**
According to https://www.libvips.org/API/current/libvips-draw.html drawing works as follows:
1. call copy
2. draw

While trying to do so I get terrible memory leaks. 

**To Reproduce**
In order to show the behavior I built a minimal (nonsense) example. In my real world application I do not call copy and draw_circle in a loop but it is actually done by the user on mouse click. I also want to copy the image before modifiying it because I push the initial image to my undo buffer in order to be able to undo changes.
```
#include <vips/vips.h>
#include <vector>
#include <vips/vips8>

using namespace std;
using namespace vips;

int main(int argc, char** argv)
{
    VIPS_INIT(argv[0]);
    VImage paintLayer = VImage::new_from_file(argv[1]);
    vector<double> ink;
    ink.push_back(127);
    for (int n = 0; n < 100; ++n)
    {
        paintLayer = paintLayer.copy();
        paintLayer.draw_circle(ink, 10, 10, 10);
    }

    return 0;
}
```
RAM diagnose output of Visual Studio
![grafik](https://user-images.githubusercontent.com/28649121/159326913-1ad8ed41-4469-49d7-afa5-ccd9292505f5.png)
I.e. the example needs more than 10GB of RAM after the loop.

**Expected behavior**
I expect the code to not produce any memory leaks.

**Actual behavior**
Look at the diagram.

**Environment**
 - OS: Windows 10
 - Vips: 8.12

**Comment by jcupitt** [comment, marked as answer]
Hi @MathemanFlo, 

`image = image.copy()` adds a null "copy" node to a pipeline. Pixels are copied between the two images, but they will have different metadata, so you can modify one without changing the other. If you do 100 of these in a row, you'll have a pipeline 100 nodes long (!!).

If you call `.draw_circle()` on a pipeline, it'll allocate a memory area, render the image into it, then draw the circle. This means that every image on your 100 node pipeline will have a separate memory copy of your source image (!!).

I think what you need here is `.copy_memory()`. This allocates ram, renders the image into it, and returns a new image which wraps that memory area. Because it breaks the pipeline, you don't get all the previous images.

I tried:

```C++
/* compile with 
 *
 *      g++ -g -Wall draw.cc `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips.h>
#include <vector>
#include <vips/vips8>

using namespace std;
using namespace vips;

int main(int argc, char** argv)
{
    VIPS_INIT(argv[0]);

    VImage paintLayer = VImage::new_from_file(argv[1]);

    vector<double> ink;
    ink.push_back(127);

    for (int n = 0; n < 100000; ++n)
    {
        paintLayer = paintLayer.copy_memory();

        paintLayer.draw_circle(ink, 10, 10, 10);
    }

    return 0;
}
```

And it runs in a steady 50mb on this PC.

**Reply by MathemanFlo** [reply]
Thank you very much for this quick and high quality answer, without you I would have had a very hard time. I learned a lot and understand the pipeline mechanism better now! Using `copy_memory` does indeed solve the memory consumption issue and also my undo-redo mechanism works like a charm now (It worked before but only due to the fact that my pipeline still contained to the old image). 
I believe copy and copy_memory and their action on the pipline should be explained in the documentation of the drawing functions. Also https://www.libvips.org/API/current/VipsImage.html#vips-image-copy-memory and https://www.libvips.org/API/8.11/cpp/classVImage.html#a852b430469f8b63648cd8c006314cd72 are so differently documented and the things you wrote clarify so much about the action on the pipeline that I could not read in the documentation. At least for me this was not obvious.

**Reply by jcupitt** [reply]
Glad it's working!

I tried to improve the docs, is this better?

https://github.com/libvips/libvips/commit/a77519f7c3a138f8b2cba635d6e337802acdc3d4

**Reply by MathemanFlo** [reply]
It is much much better now. Thank you so much! Especially the explanation which function has to be used in which situation. Maybe https://www.libvips.org/API/current/libvips-draw.html could use a reference to `copy_memory`, too. Probably this is the starter for most developers. This page has already a short explanation on the pipeline but `copy_memory` is not mentioned at all and the effect of both functions on the pipeline is also not stated explicitely. Maybe it is an overkill but I think there can never be enough examples and explanations.

**Reply by jcupitt** [reply]
I changed the advice on draw operations to use copy_memory rather than copy.

https://github.com/libvips/libvips/commit/1b97b52318e7743c7ac3803d83d8cad148e82208

copy will work, but it won't break the pipeline, so the image will include all previous images. If you're doing a lot of drawing, copy_memory is obviously better.

**Reply by MathemanFlo** [reply]
Thank you for the improvements and for your hard work on libvips! Love using it.

**Reply by jcupitt** [reply]
Oh no problem. Doc complaints are very useful: after working on something for a while it all seems so obvious it's impossible to know what bits of the docs are unclear.

**Reply by MathemanFlo** [reply]
The documentation itself is more or less sufficient. But to be honest sometimes a few working examples would be great. For example in this case if I would have had a small example how to modify an image using the draw functions then I wouldn't have had the problem. When I started using libvips I had problems with the fact that it sometimes changes the interpretation and the datatype. While it is useful and logical it was not clear to me how to convert it back to uchar or something else. An example would have been great. It was also not so easy to create new images like a mask for the morphological operations. These are not loaded from file (at least I do not do it) and it was not straight forward to create a rectanglar or circular mask. In this case an example would have been great that shows how to use the morphological functions. But in most cases the documentation is absolutely sufficient.

**Reply by jcupitt** [reply]
I tried adding some more notes to the morph docs and linking to some other mask creator functions: https://github.com/libvips/libvips/commit/75e6906f52fc66a9053703f7c51a8d6062507581

I'm not sure where to add more notes about interpretation and format. Any suggestions?

**Reply by jcupitt** [reply]
... I added some notes to the arithmetic section: https://github.com/libvips/libvips/commit/62fee01f3a621035608bb2461a28d49d2cd1c4cb

**Reply by MathemanFlo** [reply]
Thank you for all the notes. I'm sure it will help lots of people. Love the notes on morph operations because I also learned something new. The notes for interpretation and cast are excellent and definitely on point. I also thought about it yesterday and maybe a usuful place would be https://www.libvips.org/API/current/VipsImage.html. Before using libvips I had no prior experience with glib and hence it was not clear to me how to access the interpretation and all the other properties. I think it is also non obvious how to set them when reading this doc page, i.e. that copy can be used. If the user does not use the plain c interface but the c++ (or other ones) there is a member function called `interpretation()` but there is no obvious setter.

I also thought whether users would benefit from small code examples, i.e. short programms that illustrate some functionality. I'm sure there are lots of small programs scattered around that show something interesting which could be put in a directory. At least I love short working examples I can adapt to my needs.
